File: ruby/binary/binary.rb
class Binary
  attr_reader :input

  def initialize(input)
    if not_binary(input)
      @input = ['0']
    else
      @input = input.chars
    end
  end

  def to_decimal
    input.reverse.each_with_index.inject(0) do |sum, (value, i)|
      sum + value.to_i * (2 ** i)
    end
  end

  def not_binary(input)
    /[^01]/.match(input)
  end
end

File: javascript/binary/binary.js
var isBinary = function(input) {
  return input.replace(/[01]/g, '') === '';
};

var Binary = function(input) {
  if(isBinary(input)) {
    this.value = input.split('').map(Number);
  } else {
    this.value = [0];
  }
};

Binary.prototype.toDecimal = function() {
  reverseInput = this.value.reverse();
  return reverseInput.reduce(function(previousValue, currentValue, index) {
    return previousValue + currentValue * Math.pow(2, index);
  });
};

module.exports = Binary;
